### PR TITLE
Fix ProofRule::STRING_DECOMPOSE and add to ALF signature.

### PR DESCRIPTION
**Pull request by ajreynol**
This is an infrequently used rule for handling string disequalities.

It was not covered in the nightlies (since proof elaboration in the strings solver failed), and moreover its internal proof checker was bogus.

Note that the premise of the lemma was expected to be GEQ, not GT.  This weakens the premise of the lemma sent in the core solver, which matches the proof rule (and thus is sound).

This fixes these issues and adds its definition to ALF.

### CODE REVIEW

File: proofs/alf/cvc5/rules/Strings.smt3 [hunk 1]
```diff
@@ -188,6 +188,28 @@
...
+;   A conjunction stating that s is equal to two purification skolems
+;   corresponding to decomposing s at position n, and that the decomposed term
+;   has length n.
+(declare-rule string_decompose ((U Type) (s (Seq U)) (n Int) (b Bool))
+  :premises ((>= (str.len s) n))
+  :args (b)
+  :conclusion (alf.ite b
+                (let ((kp (@purify (skolem_prefix s n)))
+                      (ks (@purify (skolem_suffix_rem s n))))
+                (and (= s (str.++ kp ks)) (= (str.len ks) n)))
```
Comment (abdoo8080) [inline proofs/alf/cvc5/rules/Strings.smt3:207]:
```suggestion
                (and (= s (str.++ kp ks)) (= (str.len kp) n)))
```

File: src/theory/strings/infer_proof_cons.cpp [hunk 2]
```diff
@@ -652,6 +652,8 @@ void InferProofCons::convert(InferenceId infer,
         }
         else
         {
+          Assert(false)
+              << "Failed to convert length " << lenReq << " " << ps.d_children;
```
Comment (abdoo8080) [inline src/theory/strings/infer_proof_cons.cpp:656]: Intentional? How about `Unreachable`?

Comment (ajreynol) [reply src/theory/strings/infer_proof_cons.cpp:656]: This should only fail in debug.

The consequence of this failure is that a trusted step will be added to the proof. It is hard to ensure this elaboration always succeeds.

File: include/cvc5/cvc5_proof_rule.h [hunk 3]
```diff
@@ -1585,15 +1585,17 @@ enum ENUM(ProofRule) : uint32_t
    *   \inferrule{\mathit{len}(t) \geq n\mid \bot}{t = w_1\cdot w_2 \wedge
    *   \mathit{len}(w_1) = n}
    *
-   * or alternatively for the reverse:
+   * where :math:`w_1` is :math:`\mathit{skolem}(\mathit{pre}(t,n)` and
+   * :math:`w_2` is :math:`\mathit{skolem}(\mathit{suf}(t,n)`.
+   * Or alternatively for the reverse:
```
[the remark this answers is not in the document]
Comment (ajreynol) [reply include/cvc5/cvc5_proof_rule.h:1590]: Yes, me too. Let's revisit once all PRs are merged. I have this in my todo list.

**Review by abdoo8080 — APPROVED** [review]

**Pull request merged**